### PR TITLE
Docker Search for ISV repos [docker-search]

### PR DESCRIPTION
**Pull request by aprajshekhar**
Solr search required changes so that ISV images can be made part of search results.

### CODE REVIEW

File: crane/search/solr.py [hunk 1]
```diff
@@ -44,7 +44,6 @@ def search(self, query):
         results = self._parse(body)
         filtered_results = itertools.ifilter(self._filter_result, results)
         return itertools.imap(self._format_result, filtered_results)
-
```
Comment (mhrivnak) [inline crane/search/solr.py]: This blank line needs to be here for pep8

File: crane/search/base.py [hunk 2]
```diff
@@ -108,6 +108,6 @@ def _get_data(url):
 
 # this data structure should be used to return search results in a uniform
 # and well-defined way.
-class SearchResult(namedtuple('SearchResult', ['name', 'description', 'is_trusted',
+class SearchResult(namedtuple('SearchResult', ['name', 'description','should_filter', 'is_trusted',
```
Comment (mhrivnak) [inline crane/search/base.py]: space after the comma for pep8

File: crane/search/base.py [hunk 3]
```diff
@@ -108,6 +108,6 @@ def _get_data(url):
...
                                                'is_official', 'star_count'])):
-    result_defaults = {'is_trusted': False, 'is_official': False, 'star_count': 0}
+    result_defaults = {'should_filter':True, 'is_trusted': False, 'is_official': False, 'star_count': 0}
```
Comment (mhrivnak) [inline crane/search/base.py]: space before ``True`` for pep8

File: tests/search/test_base.py [hunk 4]
```diff
@@ -31,13 +31,14 @@ def test_format_result(self):
...
             'description': 'Red Hat Enterprise Linux',
+            'should_filter':True,
             'is_trusted': False,
             'is_official': False,
             'star_count': 0,
         })
 
     def test_non_defaults(self):
-        result = base.SearchResult('rhel', 'Red Hat Enterprise Linux',
+        result = base.SearchResult('rhel', 'Red Hat Enterprise Linux',False,
```
Comment (mhrivnak) [inline tests/search/test_base.py]: need a space before ``False`` for pep8

File: crane/search/solr.py [hunk 5]
```diff
@@ -58,13 +57,33 @@ def _parse(self, body):
...
+                    continue
+                description = item.get('ir_description', item.get('abstract'))
+                name = item.get('c_pull_command', item.get('allTitle'))
+                should_filter = True if item.get('documentKind') == 'ImageRepository' else False
                 trusted = item.get('ir_automated', SearchResult.result_defaults['is_trusted'])
                 automated = item.get('ir_official', SearchResult.result_defaults['is_official'])
                 stars = item.get('ir_stars', SearchResult.result_defaults['star_count'])
-                yield SearchResult(item['allTitle'], item['ir_description'],
-                                   trusted, automated, stars)
+                yield SearchResult(name, description, should_filter,trusted, automated, stars)
```
Comment (mhrivnak) [inline crane/search/solr.py]: need a space before ``trusted`` for pep8

File: tests/search/test_base.py [hunk 6]
```diff
@@ -45,6 +46,7 @@ def test_non_defaults(self):
         self.assertDictEqual(ret, {
             'name': 'rhel',
             'description': 'Red Hat Enterprise Linux',
+            'should_filter':False,
```
Comment (mhrivnak) [inline tests/search/test_base.py]: space before ``False`` for pep8

File: tests/search/test_gsa.py [hunk 7]
```diff
@@ -45,11 +45,12 @@ def test_workflow_filter_true(self, mock_parse_xml, mock_get_data, mock_filter):
                                                     **SearchResult.result_defaults)]
 
         ret = self.gsa.search('foo')
-
+        #should_filter is part of default
```
Comment (mhrivnak) [inline tests/search/test_gsa.py]: space after the ``#`` for pep8

File: crane/search/base.py (hunk 2 above)
Comment (mhrivnak) [inline crane/search/base.py]: Since ``should_filter`` is a new parameter, and since it is a different type of data (it's not part of the result, but rather metadata about the result), I would make it the last parameter in this data structure.

File: crane/search/solr.py [hunk 8]
```diff
@@ -58,13 +57,33 @@ def _parse(self, body):
...
+        repository, authorize it otherwise skip authorization
+
+        :param result:  one search result
+        :type  result:  SearchResult
+        :return:    True if either the repository is known and the user is authorized or
+                    if authorization can be skipped else False
+        :rtype:     bool
+        """
+        if result.should_filter:
+            return HTTPBackend._filter_result(self, result)
```
Comment (mhrivnak) [inline crane/search/solr.py]: Unless you have a reason in mind not to do so, this is a good place to use ``super``.

File: tests/search/test_base.py [hunk 9]
```diff
@@ -31,13 +31,14 @@ def test_format_result(self):
         self.assertDictEqual(ret, {
             'name': 'rhel',
             'description': 'Red Hat Enterprise Linux',
+            'should_filter':True,
```
Comment (mhrivnak) [inline tests/search/test_base.py]: need a space before ``True`` for pep8

File: tests/search/test_solr.py [hunk 10]
```diff
@@ -10,6 +10,7 @@
 
 
 class BaseSolrTest(unittest2.TestCase):
+
```
Comment (mhrivnak) [inline tests/search/test_solr.py]: I'm not picky about this, and I don't think pep8 takes a stance on it, but I do like consistency. The existing style in this module and others is to not use a blank line after a class definition, so I think we should stick with that convention.

File: tests/search/test_solr.py [hunk 11]
```diff
@@ -32,11 +35,13 @@ def test_quotes_query(self, mock_parse):
     @mock.patch('crane.search.solr.Solr._get_data', spec_set=True)
     @mock.patch('crane.search.solr.Solr._parse')
     def test_workflow_filter_true(self, mock_parse, mock_get_data, mock_filter):
-        mock_parse.return_value = [SearchResult('rhel', 'Red Hat Enterprise Linux')]
+        mock_parse.return_value = [
```
Comment (mhrivnak) [inline tests/search/test_solr.py]: Here, line 43, and line 62 seem to be unnecessarily re-formatted. The changes don't seem to be related to the functionality you're trying to add.

File: tests/search/test_solr.py [hunk 12]
```diff
@@ -54,11 +59,66 @@ def test_workflow_filter_true(self, mock_parse, mock_get_data, mock_filter):
...
+            httplib.NOT_FOUND)
+        result = SearchResult(
+            'rhel', 'Red Hat Enterprise Linux', False, False, False, 0)
+        ret_val = self.solr._filter_result(result)
+        self.assertEquals(ret_val, True)
+
+    @mock.patch('crane.app_util.repo_is_authorized', spec_set=True)
+    def test_workflow_filter_true_with_image_repository_document_kind_with_defaults(self, mock_is_authorized):
+        """
+        When the should_filter attribute of SearchResult instance is default which is Flase,
```
Comment (mhrivnak) [inline tests/search/test_solr.py]: s/Flase/False/

**Comment by mhrivnak** [comment]
Thanks for the contribution! I'm not sure why flake8 (which gets run by travis) didn't catch all of these pep8 anomalies. Sorry about that.

Aside from the comments above, we need two things:

1. Please add yourself to the AUTHORS file
2. Please squash to one commit. I usually do ``git rebase -i master`` and ``git push -f origin mybranch``

**Comment by aprajshekhar** [comment]
Thanks for reviewing my changes. I have incorporated the review comments. And added myself to Authors file as well as squashed into one commit.

File: tests/search/test_gsa.py [hunk 13]
```diff
@@ -45,14 +45,15 @@ def test_workflow_filter_true(self, mock_parse_xml, mock_get_data, mock_filter):
                                                     **SearchResult.result_defaults)]
 
         ret = self.gsa.search('foo')
-
+        # should_filter is part of default
         mock_get_data.assert_called_once_with('http://pulpproject.org/search?q=foo')
         self.assertDictEqual(list(ret)[0], {
             'name': 'rhel',
-            'description': 'Red Hat Enterprise Linux',
+            'description': 'Red Hat Enterprise Linux',
```
Comment (mhrivnak) [inline tests/search/test_gsa.py]: please remove the trailing white space

File: tests/search/test_solr.py [hunk 14]
```diff
@@ -17,11 +17,13 @@ def setUp(self):
 
 
 class TestInit(BaseSolrTest):
+
```
Comment (mhrivnak) [inline tests/search/test_solr.py]: please remove this blank line

File: tests/search/test_solr.py [hunk 15]
```diff
@@ -17,11 +17,13 @@ def setUp(self):
...
     def test_stores_data(self):
         self.assertEqual(self.solr.url_template, self.url)
 
 
 class TestSearch(BaseSolrTest):
+
```
Comment (mhrivnak) [inline tests/search/test_solr.py]: please remove this blank line

**Comment by mhrivnak** [comment]
Thanks! There are four more comments to address; just little formatting things. Then we're ready to merge.

File: tests/search/test_solr.py [hunk 16]
```diff
@@ -49,14 +50,66 @@ def test_workflow_filter_true(self, mock_parse, mock_get_data, mock_filter):
     @mock.patch('crane.search.solr.Solr._get_data', spec_set=True)
     @mock.patch('crane.search.solr.Solr._parse')
     def test_workflow_filter_true(self, mock_parse, mock_get_data, mock_filter):
-        mock_parse.return_value = [SearchResult('rhel', 'Red Hat Enterprise Linux',
-                                                **SearchResult.result_defaults)]
+        mock_parse.return_value = [SearchResult('rhel', 'Red Hat Enterprise Linux', **SearchResult.result_defaults)]
```
Comment (mhrivnak) [inline tests/search/test_solr.py]: did this cross the 100 character limit?

**Comment by aprajshekhar** [comment]
Made the required changes. Please review.

File: tests/search/test_solr.py (hunk 10 above)
Comment (rbarlow) [inline tests/search/test_solr.py]: On 06/17/2015 05:08 PM, Michael Hrivnak wrote:
> I'm not picky about this, and I don't think pep8 takes a stance on it,
> but I do like consistency. The existing style in this module and others
> is to not use a blank line after a class definition, so I think we
> should stick with that convention.

One day far into the future, I'd like to start using the pep257 module
that is used by a few other projects (like pulp_python and
pulp_openstack). PEP-257 apparently does want this blank line, at least
the pep257 checker complains without it.

Of course, we can add exceptions for the rules we don't care about, and
this does seem fairly arbitrary to me. My personal preference is also
not to have a blank line after a class definition, but to immediately
jump into the docblock.

The great thing about pep257 is that it can enforce the existence of doc
blocks (and I think param definitions too?)

-- 
Randy Barlow

File: AUTHORS [hunk 17]
```diff
@@ -2,4 +2,4 @@ Barnaby Court (bcourt@redhat.com)
 Michael Hrivnak (mhrivnak@redhat.com)
 Sayli Karmarkar (skarmark@redhat.com)
 Dennis Kliban (dkliban@redhat.com)
-Aaron Weitekamp (aweiteka@redhat.com)
```
Comment (mhrivnak) [inline AUTHORS]: I doubt you meant to do this. Please restore Aaron's name. :)

Comment (aprajshekhar) [inline AUTHORS]: How did I miss that? :( Sorry about that. Restored.

File: AUTHORS [hunk 18]
```diff
@@ -2,4 +2,5 @@ Barnaby Court (bcourt@redhat.com)
...
+Aaron Weitekamp (aweiteka@redhat.com
```
Comment (mhrivnak) [inline AUTHORS]: Sorry to be a perfectionist, but could you add the trailing ``)`` ? Thanks! Then I promise I'll merge it.
